crates/kubelet/src/device_plugin_manager/mod.rs

```diff
@@ -0,0 +1,954 @@
+                                }
+                                update_node_status = true;
+                            }
+                        });
+
+                        // (3) Check if Device removed
+                        previous_endpoint_devices
+                            .iter()
+                            .for_each(|(_, previous_device)| {
+                                if !response.devices.contains(previous_device) {
+                                    // TODO: how to handle already allocated devices? Pretty sure K8s lets them keep running but what about the allocated_device map?
+                                    all_devices
+                                        .lock()
+                                        .unwrap()
+                                        .get_mut(&list_and_watch_resource_name)
+                                        .unwrap()
+                                        .remove(&previous_device.id);
+                                    update_node_status = true;
+                                }
+                            });
+
+                        // Replace previous devices with current devices
+                        previous_endpoint_devices = current_devices;
+
+                        if update_node_status {
+                            // TODO handle error -- maybe channel is full
+                            update_node_status_sender.send(()).unwrap();
+                        }
+                    }
+                    // TODO: remove endpoint from map
+                }
+            }
+        });
+
+        // Only add device plugin to map if successful ListAndWatch call
+        if successful_connection_receiver.await.unwrap() == success {
+            let endpoint = Endpoint {
+                client,
+                register_request: register_request.clone(),
+            };
+            self.add_plugin(endpoint);
+        } else {
+            return Err(anyhow::Error::msg(format!(
+                "could not call ListAndWatch on device plugin at socket {:?}",
+                register_request.endpoint
+            )));
+        }
+
+        Ok(())
+    }
+
+    /// This is the call that you can use to allocate a set of devices
+    /// from the registered device plugins.
+    /// Takes in a map of devices requested by containers, keyed by container name.
+    pub async fn do_allocate(
+        &self,
+        pod: &Pod,
+        container_devices: HashMap<String, HashMap<String, Quantity>>,
+    ) -> anyhow::Result<()> {
+        let mut all_allocate_requests: HashMap<String, Vec<ContainerAllocateInfo>> = HashMap::new();
+        let mut updated_allocated_devices = false;
+        for (container_name, requested_resources) in container_devices {
+            for (resource_name, quantity) in requested_resources {
+                let num_requested: usize =
+                    serde_json::to_string(&quantity).unwrap().parse().unwrap();
```

Comment:
Good point! CPU and memory resource requests, for example, allow quantities to be requested in `100m` and `500MI` units which would cause this to panic. In b1f3a819d19, I broke this out so that the quantity is only parsed after checking that the resource is a device plugin resource, since device plugin resources should be requested in numerical quantities. If this is not the case, then I bubble up the error (instead of unwrapping and panicking). Does this seem better?

Reply:
Yeah, that looks better